Lock component capable to open door latches

PH89 opened the pull request:
**Rework lock component:**
- The lock component is now a own component that is not longer using the standard toggle component. 
- The lock component now supports opening doors for compatible locks (like the HomeMatic KeyMatic device).
- The lock component now uses separate buttons that are indicating the status of the lock component.

PR in main project: https://github.com/home-assistant/home-assistant/pull/11980

PH89 committed:
Improves lock component.

Lock component is now not longer shared with a standard toggle component. The lock component now supports opening doors for compatible locks (like the HomeMatic KeyMatic device).

Instead of using a toggle button the component now uses separate buttons that are indicating the status of the lock component.

PH89 committed:
Merge branch 'master' of https://github.com/PH89/home-assistant-polymer

homeassistant commented:
Hi @PH89, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant-polymer#845).

Once you do that we will be able to review and accept this pull request.

Thanks!

PH89 committed:
travis improvements.

PH89 committed:
travis improvements.

c727 commented on src/state-summary/state-card-lock.html:
```diff
@@ -0,0 +1,42 @@
+
+<dom-module id="state-card-lock">
+  <template>
+    <style is="custom-style" include="iron-flex iron-flex-alignment"></style>
```
is="custom-style" is old syntax and can be removed

PH89 replied:
This was just a clone from state-card-toggle. I removed it.

andrey-git commented on src/components/entity/ha-entity-lock.html:
```diff
@@ -0,0 +1,154 @@
+      }
+    </style>
+
+    <template is='dom-if' if='[[!stateObj.attributes.assumed_state]]'>
```
In "assumed_state" mode you should still show the buttons, but without deciding which one is active

PH89 replied:
Yes this can be a problem. I will provide improvments.

andrey-git commented on src/components/entity/ha-entity-lock.html:
```diff
@@ -0,0 +1,154 @@
+          observer: 'stateObjObserver'
+        },
+
+        toggleChecked: {
```
You copied this from toggle. Is it needed here?

PH89 replied:
Its needed, since it is still a toggle state model.

andrey-git reviewed (commented):
Could you attach a screenshot of the resulting UI?

PH89 commented:
This is the new UI look.
First 2 rows are demo lock
Last row is the KeyMatic lock or every other lock that is capable to open the door latch.

![newlockui](https://user-images.githubusercontent.com/11292780/35482916-d5bfc4c2-043b-11e8-88f0-9782e5a84dca.PNG)

PH89 committed:
Improvements from code review.

* removing custom-style attribute
* removing template tag since buttons should be always present, even if th state is assumed.
* removing unneeded function toogleChange

andrey-git commented:
I think the UI shouldn't contain extra icons. So if we know the state (i.e. !assumed_state) show just a single lock/unlock icon and and "open" icon if needed.

Only show both icons in case of assumed_state.

PH89 commented:
@andrey-git Actually I prefer the extra icon in this case. Because a lock is a very important device, the user should clearly see its state (blue highlight) and should choose explicit if he/she wants to lock/unlock.

If you only have two icons its much harder for the user to determine if he/she is unlocking or locking.

I would like to quote the zen of python by Tim Peters here:

> Explicit is better than implicit.

:-)

andrey-git commented:
This is (material design) UI which very different from Python :)

Lets defer to @balloob on this.

andrey-git commented on src/state-summary/state-card-lock.html:
```diff
@@ -0,0 +1,42 @@
+    <style>
+      ha-entity-lock {
+        margin: -4px -16px -4px 0;
+        padding: 4px 16px;
```
This was used for toggle to increase the tap target to 48px.
Since ha-entity-lock doesn't have a single tap target - this is not needed

PH89 committed:
Removing unneeded stying block

PH89 commented:
@andrey-git I implemented your suggestion. Do you still think that this looks better?

![newlockui2](https://user-images.githubusercontent.com/11292780/35485069-2e45459a-045a-11e8-9d24-fcc7e75ad026.PNG)

c727 commented:
much better, maybe move the open door to the left. Or maybe open door should be text? idk

c727 commented:
2nd question: why don't we use the standard on/off toggle here?

andrey-git commented:
Keeping the "open" to the left will make it more aligned.

I'm undecided on whether I like this or the toggle better.

PH89 commented:
@c727 1. So you want icons but just the unlock button as text ? Its a break in design in my opinion.
2. Its not a toggle for a reason. Because a lock is not a switch. Its not clear for a user what will happend if the toogle is on / off state. Maybe for you is the on-state the lock is active, so it is locked. Some other guy maybe says okay off-state is locked state because the door is not usable anymore.

General I really respect your material design principles. I was taking a look into it the guildlines of it also. I couldn't find any section that says its forbidden to have three button in a row on a card. 

But even if it would say its forbidde, in some cases its okay to break the principles, since useablity matters and not the design. Especilly in a security related feature.

andrey-git commented on src/components/entity/ha-entity-lock.html:
```diff
@@ -0,0 +1,142 @@
+      }
+    </style>
+
+    <paper-button-group>
```
I'm not sure you need paper-button-group here. If you do - please import it.

PH89 replied:
Yes I removed it already.

PH89 committed:
Design more minimal, adds confirmation dialog

According to the reviewers the design is now a more minimal.
Also introduces an confirmation dialog for the "open door latch"-feature,
that should prevent the user from triggering it unintentionally.

PH89 committed:
fixes styling problems on confirmation dialog on small screens.

PH89 commented:
Current design of card:
![currentstate](https://user-images.githubusercontent.com/11292780/36064645-6c6e1ac0-0e8e-11e8-9aab-127f9fa5fdf1.PNG)
New confirmation dialog after clicking on open door latch button:
![dialog](https://user-images.githubusercontent.com/11292780/36064646-6c84d71a-0e8e-11e8-974f-762d20fa4efa.png)

PH89 commented:
I extended also the demo component on backend so you can test better
See https://github.com/home-assistant/home-assistant/pull/11980/commits/460240d2161b10278f8d5199573c939e3dba8ec7

c727 commented:
I think you should split the OK-dialog to a separate PR (Imo it's useful for config manager and hassio panel)

I suggest to move the `open door` button to the more info card, without the OK-dialog, as it's a critical function (design like a script state card: `Open door    OPEN`.

c727 commented:
... a 2nd reason to move it to more info card is that it's a different type of action (for cover start_open/close and stop)

I'm not sure if my suggestion is the best solution here, as we have different types of state cards:
```
sensors: static icon, state without actions
toggles: static icon, state-action-combo
covers: state in icon, (static) action buttons

lock: state in icon, (dynamic) action buttons
```

Maybe someone else can say his opinion, @balloob

PH89 commented:
@c727 I can move the OK-dialog to another PR. But I don't see what difference this makes. Even if this component is useful for other components as well, currently only the new lock component is using it. So I guess it can also stay here till this finally gets merged...

c727 commented:
well my intentions was to help you to get this merged and with the example above I wanted to show that your PR is different from the covers domain ( - something new: 2 action buttons in 1 card with different types of action(lock/unlock & open door).
I suggested to move it to more info card as other components also moved secondary actions to more info

PH89 commented:
@c727 Its not a secondary action imho. A lock can be locked, unlocked and open. 
Why this is different from covers? A cover can be going up / down and can be stopped. Why not move stop button into the more info card? :man_facepalming: 

Opening the door latch is a very common use case, often more important then the unlock feature, since doors with just a knob on the outside CAN'T be opened without pulling the door latch. 
Therefore this button should be in front card WITH an confirmation dialog.

General I was also not happy that you wanted me to reducing the three buttons that I did in first place. I still prefer the three button solution and so do all my colleagues and friends that I showed this feature, but I thought okay you can live also with the reduced button set also. 

But now, with your last suggestions, I have more and more the feeling that your changes requests are just going into the direction what is your personal taste and not was is improving the usability of this feature or what is a technical problem and needs really needs to be changed.

I think if many people are working on such a project, then you need to find compromises on that points that are affecting your personal tastes... This is my first contribution to this project and I am motivated to submit more and having good ideas... But such reviews are taking my moral... :disappointed: 


Hopefully some owner can solve this dispute...

c727 commented:
my last statement to this PR: Imo we discussed stuff and collected different suggestions and don't have a dispute :princess:

gohassgo commented:
I like the look of one icon. That way it works like every other switch, light, cover, etc. It would also be great if the icon would change colors like light icons do and be configurable in themes.

PH89 commented:
@gohassgo For two states components this may fits, but what about the door latch feature? I think it can’t be controlled with just one button. Highlight icons is surely possible you have ideas for colors?

gohassgo commented:
I guess I'm not familiar with latch type devices (sorry!). In the latch case it does make sense to have the door icon as well, since the lock is capable of an action with the door.

Come to think of it, it would be nice to be able to combine a lock and binary door sensor into one line like this--more compact and clean, seeing all you need to know about a door and it's lock's status. But that's not for this pr :)

As far as colors, I'd prefer the default (black?) for closed or locked and red for open or unlocked. That's what I use for templates for motion sensors. But I'm sure there are some who would prefer emphasizing the opposite, like green for locked and black for unlocked, or both with green for locked and red for unlocked. Would colors be configurable in customizer or with themes? I think if they're not, not everyone will be happy with whatever colors are used, including black/black ;)

fanaticDavid commented:
Just my 2 cents: maybe the `door` device class for binary sensors can be used as inspiration?
When the door is open, the icon on the left lights up (yellow(-ish) in the default theme) and the mdi icon `door-open` is shown.
When the door is closed, the icon on the left doesn't light up (blue(-ish) in the default theme) and the mdi icon `door-closed` is shown.

NovapaX commented:
This is confusing. I expect to see the state of the lock on the right. Therefore I liked your first implementation.
Or it should be a toggle with icons, but that would get a bit messy.

And isn’t there an icon where you can actually clearly see the difference between locked and unlocked without putting it under a microscope? Because with these icons it’s better to use text imo.

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.

PH89 commented:
@balloob This PR is waiting for a decision of a project manager regarding the design... Maybe you are willing to make one?

balloob commented:
Oh. 

So let me start with the things I agree with:
 - Toggle is unsuitable, especially for something as important as locking/unlocking
 - The lock/unlock icons are only a few pixels different, making it difficult to distinguish

I like the idea of having text and only showing the options that are applicable. Obvious always wins.

OPEN, UNLOCK, LOCK.

It is quite long. But really only in the case of an assumed lock that can open. Are those common?

PH89 commented:
@balloob LGTM I will implement it with text buttons then.
Any comment to the modal confirmation dialog? Can it stay here or should I move it to a separate PR?

balloob commented on src/home-assistant.html:
```diff
@@ -274,6 +277,11 @@
 
+  handleConfirmation(ev) {
+    ev.stopPropagation();
+    this._updateHass({ confirmationObj: ev.detail.confirmationObj });
```
We should just set `confirmationObj: ev.detail`

balloob commented:
Do we really need the confirmation dialog? Why wouldn't we use `window.confirm()` ? Browsers have made it look a lot better and especially on phones it is a lot better experience.

c727 commented:
Closing this for inactivity so someone else can work on a solution
